huaweicloud/services/vpn/resource_huaweicloud_vpn_connection.go

```diff
@@ -0,0 +1,972 @@
+// ---------------------------------------------------------------
+// *** AUTO GENERATED CODE ***
+// @Product VPN
+// ---------------------------------------------------------------
+
+package vpn
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"regexp"
+	"strings"
+	"time"
+
+	"github.com/chnsz/golangsdk"
+
+	"github.com/hashicorp/go-multierror"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/resource"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/validation"
+	"github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/common"
+	"github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/config"
+	"github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/utils"
+	"github.com/jmespath/go-jmespath"
+)
+
+func ResourceConnection() *schema.Resource {
+	return &schema.Resource{
+		CreateContext: resourceConnectionCreate,
+		UpdateContext: resourceConnectionUpdate,
+		ReadContext:   resourceConnectionRead,
+		DeleteContext: resourceConnectionDelete,
+		Importer: &schema.ResourceImporter{
+			StateContext: schema.ImportStatePassthroughContext,
+		},
+		Timeouts: &schema.ResourceTimeout{
+			Create: schema.DefaultTimeout(10 * time.Minute),
+			Update: schema.DefaultTimeout(10 * time.Minute),
+			Delete: schema.DefaultTimeout(10 * time.Minute),
+		},
+
+		Schema: map[string]*schema.Schema{
+			"region": {
+				Type:     schema.TypeString,
+				Optional: true,
+				Computed: true,
+				ForceNew: true,
+			},
+			"name": {
+				Type:        schema.TypeString,
+				Required:    true,
+				Description: `The name of the VPN connection.`,
+				ValidateFunc: validation.All(
+					validation.StringMatch(regexp.MustCompile(`^[A-Za-z0-9_-]+$`),
+						"the input is invalid"),
+					validation.StringLenBetween(0, 64),
+				),
+			},
+			"gateway_id": {
+				Type:         schema.TypeString,
+				Required:     true,
+				ForceNew:     true,
+				Description:  `The VPN gateway ID.`,
+				ValidateFunc: validation.StringLenBetween(0, 64),
+			},
+			"gateway_ip": {
+				Type:         schema.TypeString,
+				Required:     true,
+				ForceNew:     true,
+				Description:  `The VPN gateway IP address.`,
+				ValidateFunc: validation.StringLenBetween(0, 64),
+			},
+			"vpn_type": {
+				Type:        schema.TypeString,
+				Required:    true,
+				ForceNew:    true,
+				Description: `The connection type. The value can **policy**, **static** and **bgp**.`,
```

Comment:
s/can/can be.  does this auto generate? seems different with the docs, should we update the orign yaml instead?

Reply:
fixed